### ISSUE TITLE
Semantics of errors thrown in tasks and nested tasks

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19447]
Background: According to [the spec](https://chapel-lang.org/docs/master/language/spec/error-handling.html#parallelism) and #7046, errors thrown by individual tasks of a forall/coforall/... are collated into a single `TaskErrors` error. If there are nested TaskErrors errors, their contents are flattened into the parent.

This issue asks what happens when multiple tasks of the same forall or task-parallel construct throw errors. For example:

**(all)** All tasks that need to generate an error do so. Throwing an error in one task does not affect the other tasks. For example, if none of these tasks generate TaskErrors, the number of elements in the TaskErrors error generated by the forall/task-parallel construct equals the number of its tasks.

**(some)** Out of all tasks that need to generate an error, the TaskErrors error of the enclosing forall/task-parallel construct contains at least one and at most all such errors.

Consider the following two programs:
```chpl
// a simple version
proc main throws {
  coforall x in 1..2 do
    coforall z in 1..3 do
      throw new Error("hi");
}
```
```chpl
// the same task structure created with parallel iterators
proc main throws {
  forall idx in outerIter() do
    throw new Error("hi");
}
iter outerIter(param tag) {   // standalone iterator
  coforall x in 1..2 do
    forall y in innerIter() do
      yield 0;
}
iter innerIter(param tag) {   // standalone iterator
  coforall z in 1..3 do
    yield 0;
}
iter outerIter() { yield 0; } // unused serial iterator
iter innerIter() { yield 0; } // unused serial iterator
```
**(all)**: both programs produce exactly 6 "hi" errors in the TaskErrors error that gets thrown out of main.
**(some)**: each program can produce anywhere from 1 to 6 "hi" errors.

**Comment by bradcray** [comment]
> what happens when multiple tasks of the same forall or task-parallel construct throw errors.

To me, this question is practically the answer to the question "Why does Chapel have `TaskErrors`?"  I.e., isn't the whole point of `TaskErrors` to collect tasks from disparate tasks in cases like the ones you're asking about?  Or put another way, if my implementation always only ever placed a single error in `TaskErrors`, I wouldn't need to use `TaskErrors` at all, would I?  That makes it seem as though something has been lost if an implementation did that.

My intuition would be that six errors would be collected into `TaskErrors` in your example.  In what cases would a Chapel implementation want or need to put fewer than 6 errors into it?

The only guess I've come up with is "storing all those errors could exceed memory limitations", which I suppose could happen on a very large run, say hundreds of thousands of cores, if they all threw.  So I could imagine saying "it throws all tasks subject to resource constraints", but I'd expect that case not to occur very often, since it would require all tasks to throw, which seems unlikely if we think of errors as being exceptional.

**Comment by vasslitvinov** [comment]
We discussed `TaskErrors` in #7046 and I am not questioning that here. Also, I am talking about the semantics rather than implementation.  I assume that the implementation needs to do whatever we decide here semantically.

Also I am not concerned about memory limitations. If we were to think of those, we'd have to edit the spec to say that `new` returns a new class instance subject to resource constraints, a task construct spawns tasks subject to resource constraints, a function call invokes a function subject to available stack space, etc.

**(all)** and "6 errors" are my preferred answers as well.

The alternative of "possibly fewer than 6" that @lydia-duncan and @mppf have suggested makes sense if we decide that an error thrown in one task interferes with launching other tasks in the same forall/task-parallel construct.

Let's say that a coforall is launching multiple tasks. One of the tasks throws right away. Does it mean that the coforall does not need to launch the remaining tasks?

**Comment by bradcray** [comment]
> One of the tasks throws right away. Does it mean that the coforall does not need to launch the remaining tasks?

I see.  I think it would be acceptable if an implementation were to do this, but I would also not fault an implementation for not doing it.  So I suppose that means that, semantically, I'm more on the `some` side.  But it'd be fine with me if our implementation were an `all` implementation.  If we take throwing errors as being rare / exceptional, then presumably correct code would not waste much time spawning off the tasks unnecessarily (and I could imagine it being simpler to write that way).

**Comment by vasslitvinov** [comment]
Also there is a bug in the current implementation. It is what made me raise the semantic question.

The second program in the OP throws 6 errors, however reports only 3 of them within the final `TaskErrors`. The other 3 are discarded. I suggest that once an error is thrown, it should always be accounted for in its catch-block or TaskErrors aggregator.

<details><summary>the second program in the OP</summary>

```chpl
proc main throws {
  forall idx in outerIter() do
    throw new Error("hi");
}
iter outerIter(param tag) {   // standalone iterator
  coforall x in 1..2 do
    forall y in innerIter() do
      yield 0;
}
iter innerIter(param tag) {   // standalone iterator
  coforall z in 1..3 do
    yield 0;
}
iter outerIter() { yield 0; } // unused serial iterator
iter innerIter() { yield 0; } // unused serial iterator
```
</details>

**Comment by mppf** [comment]
I agree that's a bug. We could perhaps write spec-y language that reflects we want this property like this:

When errors are thrown within a parallel construct (e.g. coforall, begin, cobegin, forall), the implementation will collect any errors that were thrown into a TaskErrors that can be caught by code invoking the parallel construct. Note however that the implementation may (but is not required to) skip running some of the tasks if another task has already thrown an error.
